File: App/modules/chess-engine/src/ChessEngineModule.ts
import { EventEmitter, NativeModule, requireNativeModule } from 'expo-modules-core'

export interface PieceInfo {
  type: string
  color: string
}

export interface SearchProgressData {
  searchId: number
  depth: number
  bestMove: string
  score: number
  nodesSearched: number
  timeMs: number
}

export interface SearchResultData {
  searchId: number
  bestMove: string
  score: number
  depthCompleted: number
  nodesSearched: number
  timedOut: boolean
  cancelled?: boolean
  totalTimeMs: number
  progressHistory: SearchProgressData[]
}

export interface ChessEngineEventsMap {
  onSearchProgress: (event: SearchProgressData) => void
  [key: string]: (...args: any[]) => void
}

export interface ChessEngineModuleInterface extends NativeModule<ChessEngineEventsMap> {
  newGame(): void
  makeMove(move: string): boolean
  undoMove(): void
  getLegalMoves(): string[]
  getLegalMovesFrom(square: string): string[]
  isMoveLegal(move: string): boolean
  getCurrentPlayer(): string
  isInCheck(): boolean
  isCheckmate(): boolean
  isStalemate(): boolean
  isDraw(): boolean
  getBoard(): PieceInfo[]
  getFEN(): string
  loadFromFEN(fen: string): boolean
  getBestMove(depth: number, maxTimeMs?: number, aiVersion?: 'v1' | 'v2'): Promise<string>
  getBestMoveAtDepth(depth: number, maxTimeMs?: number, aiVersion?: 'v1' | 'v2'): Promise<string>
  searchBestMove(
    searchId: number,
    maxDepth: number,
    maxTimeMs?: number,
    aiVersion?: 'v1' | 'v2',
  ): Promise<SearchResultData>
  getMoveHistory(): string[]
  canUndo(): boolean
  evaluatePosition(): number
  clearSearchCaches(): void
}

const ChessEngineModule = requireNativeModule<ChessEngineModuleInterface>('ChessEngine')

export type SearchProgressListener = (event: SearchProgressData) => void

export function addSearchProgressListener(listener: SearchProgressListener): { remove(): void } {
  // Native modules extend EventEmitter at runtime, cast to access event methods
  const emitter = ChessEngineModule as unknown as InstanceType<
    typeof EventEmitter<ChessEngineEventsMap>
  >
  return emitter.addListener('onSearchProgress', listener)
}

export default ChessEngineModule

File: App/modules/chess-engine/ios/ChessEngineModule.swift
import ExpoModulesCore

public class ChessEngineModule: Module {
  private var engine: ChessEngineWrapper?
  private let initLock = NSLock()
  private let engineQueue = DispatchQueue(label: "com.chessapp.engine.queue", qos: .userInitiated)

  public func definition() -> ModuleDefinition {
    Name("ChessEngine")

    Events("onSearchProgress")

    OnCreate {
      self.initLock.lock()
      defer { self.initLock.unlock() }
      if self.engine == nil {
        self.engine = ChessEngineWrapper()
      }
    }

    OnDestroy {
      self.initLock.lock()
      defer { self.initLock.unlock() }
      self.engine = nil
    }

    Function("newGame") {
      self.ensureInitialized()
      guard let engine = self.engine else { return }
      engine.clearSearchCaches()
      self.engineQueue.sync {
        engine.newGame()
      }
    }

    Function("makeMove") { (move: String) -> Bool in
      self.ensureInitialized()
      guard let engine = self.engine else { return false }
      engine.clearSearchCaches()
      return self.engineQueue.sync {
        engine.makeMove(move)
      }
    }

    Function("undoMove") {
      self.ensureInitialized()
      guard let engine = self.engine else { return }
      engine.clearSearchCaches()
      self.engineQueue.sync {
        engine.undoMove()
      }
    }

    Function("getLegalMoves") { () -> [String] in
      self.ensureInitialized()
      return self.engine?.getLegalMoves() ?? []
    }

    Function("getLegalMovesFrom") { (square: String) -> [String] in
      self.ensureInitialized()
      return self.engine?.getLegalMoves(from: square) ?? []
    }

    Function("isMoveLegal") { (move: String) -> Bool in
      self.ensureInitialized()
      return self.engine?.isMoveLegal(move) ?? false
    }

    Function("getCurrentPlayer") { () -> String in
      self.ensureInitialized()
      return self.engine?.getCurrentPlayer() ?? "white"
    }

    Function("isInCheck") { () -> Bool in
      self.ensureInitialized()
      return self.engine?.isInCheck() ?? false
    }

    Function("isCheckmate") { () -> Bool in
      self.ensureInitialized()
      return self.engine?.isCheckmate() ?? false
    }

    Function("isStalemate") { () -> Bool in
      self.ensureInitialized()
      return self.engine?.isStalemate() ?? false
    }

    Function("isDraw") { () -> Bool in
      self.ensureInitialized()
      return self.engine?.isDraw() ?? false
    }

    Function("getBoard") { () -> [[String: Any]] in
      self.ensureInitialized()
      let board = self.engine?.getBoard() ?? []
      return board.compactMap { item in
        if let dict = item as? [String: Any] {
          return dict
        }
        return [:]
      }
    }

    Function("getFEN") { () -> String in
      self.ensureInitialized()
      return self.engine?.getFEN() ?? ""
    }

    Function("loadFromFEN") { (fen: String) -> Bool in
      self.ensureInitialized()
      guard let engine = self.engine else { return false }
      engine.clearSearchCaches()
      return self.engineQueue.sync {
        engine.load(fromFEN: fen)
      }
    }

    AsyncFunction("getBestMove") { (depth: Int, maxTimeMs: Int, aiVersion: String, promise: Promise) in
      self.ensureInitialized()
      guard let engine = self.engine else {
        promise.resolve("")
        return
      }

      self.engineQueue.async {
        let move = engine.getBestMove(Int32(depth), maxTimeMs: Int32(maxTimeMs), aiVersion: aiVersion)
        promise.resolve(move)
      }
    }

    AsyncFunction("getBestMoveAtDepth") { (depth: Int, maxTimeMs: Int, aiVersion: String, promise: Promise) in
      self.ensureInitialized()
      guard let engine = self.engine else {
        promise.resolve("")
        return
      }

      self.engineQueue.async {
        let move = engine.getBestMove(atDepth: Int32(depth), maxTimeMs: Int32(maxTimeMs), aiVersion: aiVersion)
        promise.resolve(move)
      }
    }

    AsyncFunction("searchBestMove") { (searchId: Int, maxDepth: Int, maxTimeMs: Int, aiVersion: String, promise: Promise) in
      self.ensureInitialized()
      guard let engine = self.engine else {
        promise.resolve([
          "searchId": searchId,
          "bestMove": "",
          "score": 0,
          "depthCompleted": 0,
          "nodesSearched": 0,
          "timedOut": false,
          "cancelled": false,
          "totalTimeMs": 0,
          "progressHistory": []
        ])
        return
      }

      self.engineQueue.async {
        let result = engine.searchBestMove(
          Int32(searchId),
          maxDepth: Int32(maxDepth),
          maxTimeMs: Int32(maxTimeMs),
          aiVersion: aiVersion
        ) { progress in
          DispatchQueue.main.async {
            // Convert [AnyHashable : Any] to [String : Any?] for sendEvent
            let progressDict = progress.reduce(into: [String: Any?]()) { result, pair in
              if let key = pair.key as? String {
                result[key] = pair.value
              }
            }
            self.sendEvent("onSearchProgress", progressDict)
          }
        }
        promise.resolve(result)
      }
    }

    Function("getMoveHistory") { () -> [String] in
      self.ensureInitialized()
      return self.engine?.getMoveHistory() ?? []
    }

    Function("canUndo") { () -> Bool in
      self.ensureInitialized()
      return self.engine?.canUndo() ?? false
    }

    Function("evaluatePosition") { () -> Int in
      self.ensureInitialized()
      return Int(self.engine?.evaluatePosition() ?? 0)
    }

    Function("clearSearchCaches") {
      self.ensureInitialized()
      self.engine?.clearSearchCaches()
    }
  }

  private func ensureInitialized() {
    initLock.lock()
    defer { initLock.unlock() }
    if engine == nil {
      engine = ChessEngineWrapper()
    }
  }
}
